### PR TITLE
[RLlib] Attention Net prep PR #3.

### PR DESCRIPTION
**Pull request by sven1977**
The current attention net trajectory view PR (#11729) is too large (>1000 lines added).
Therefore, I'm moving smaller preparatory and cleanup changes in ~2 pre-PRs. This is the third one of these. Only review it once this 2nd one here (#12449) has been merged.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 1]
```diff
@@ -349,6 +374,16 @@ def setup_mixins(policy: Policy, obs_space: gym.spaces.Space,
     LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
 
 
+def view_requirements_fn(policy):
+    # Adds the input-dict used in postprocessing to the dict of view-reqs.
+    # This is for value calculation at the very end of a trajectory
+    # (abs_pos=-1). It's only used if the trajectory is not finished yet and we
+    # have to rely on the last value function output as a reward estimation.
+    return {
+        "_value_input_dict": ViewRequirement(is_input_dict=True, abs_pos=-1)
```
Comment (ericl) [inline rllib/agents/ppo/ppo_tf_policy.py:383]: Wondering if there are better names for this
```
# Can specify either abs_index or shift
abs_index=-1
shift=x
```

I think the notion of "index" is clearer in Python (-1 index means end). Also, we get to keep shift.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -157,14 +157,15 @@ NodeManager::NodeManager(boost::asio::io_service &io_service, const NodeID &self
...
-                            on_objects_spilled),
+      local_object_manager_(
+          RayConfig::instance().free_objects_batch_size(),
+          RayConfig::instance().free_objects_period_milliseconds(), worker_pool_,
+          gcs_client_->Objects(), worker_rpc_pool_,
+          [this](const std::vector<ObjectID> &object_ids) {
+            object_manager_.FreeObjects(object_ids,
+                                        /*local_only=*/false);
+          },
+          on_objects_spilled),
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:168]: ? revert change

File: rllib/policy/view_requirement.py [hunk 3]
```diff
@@ -48,12 +50,29 @@ def __init__(self,
...
+                a range via a str: `data_rel_pos="-100:0"`, which will pass in
+                the past 100 observations plus the current one.
+            abs_pos (Optional[int]): An optional absolute position arg,
+                used e.g. for the location of a requested inference dict within
+                the trajectory. Negative values refer to counting from the end
+                of a trajectory.
             used_for_training (bool): Whether the data will be used for
                 training. If False, the column will not be copied into the
                 final train batch.
+            is_input_dict (bool): Whether the "view" of this requirement is an
```
Comment (ericl) [inline rllib/policy/view_requirement.py:63]: This seems odd, what is the reason we need it? Are there any cleaner alternatives?

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 4]
```diff
@@ -292,25 +302,40 @@ def __init__(self, obs_space, action_space, config):
...
-                    SampleBatch.PREV_REWARDS: tf.convert_to_tensor(
-                        [prev_reward]),
-                    "is_training": tf.convert_to_tensor([False]),
-                }, [tf.convert_to_tensor([s]) for s in state],
-                                          tf.convert_to_tensor([1]))
-                # [0] = remove the batch dim.
-                return self.model.value_function()[0]
+            # Input dict is provided to us automatically via the policy-defined
+            # "view". It's a single-timestep (last one in trajectory)
+            # input_dict.
```
Comment (ericl) [inline rllib/agents/ppo/ppo_tf_policy.py:307]: What is the reason for this change? Could we get the previous code to work without adding this if branch?

**Review by ericl — CHANGES_REQUESTED** [review]
@sven1977 main question is the input dict option. Maybe I'm missing something, but it doesn't seem like a necessary change. Correct me if this is wrong.

File: rllib/policy/view_requirement.py (hunk 3 above)
Comment (sven1977) [reply rllib/policy/view_requirement.py:63]: This option will be necessary for attention nets. We shouldn't have attention net or RNN-specific code in the postprocessing fn (e.g. of PPO). Instead, it's like saying: "I need an input_dict here, provide one given the model's requirements".

File: rllib/agents/ppo/ppo_tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo_tf_policy.py:383]: Fair enough, I'll rename these two:
abs_pos -> index
data_rel_pos -> shift

Comment (sven1977) [reply rllib/agents/ppo/ppo_tf_policy.py:383]: On the input_dict:
It's not necessary for this very PR, but as this is a preparatory PR (to make the attention PR smaller) I decided to already add this here. The attention net PR needs this feature to be able to not have to boiler-plate/hardcode the attention logic inside e.g. PPO's postprocessing fn (this function should not have to worry about the model being an RNN or attention net, it should not need to know).

**Comment by ericl** [comment]
@sven1977 Can you give some examples of how it works in either way w/wo attention net? I have a hard time understanding why we need the concept of an input dict for certain models (aren't all views subclassing a dict?)

**Comment by sven1977** [comment]
a) For the input_dict that's we currently build on-the-fly in all PGs' postprocessing functions, we only need the very last timestep of the trajectory. However, we currently copy the entire trajectory's state_out and next_obs for that (even though we only use the very last timestep).
b) The policy should not have to worry about how to construct this input_dict (we currently check, whether we have an RNN, then use the last state_out as state_in, etc.. We would have to add another check for attention nets (the attention net needs more than the last state_out due to its memory input requirements). So instead, we create this view, then let the SampleCollector worry about generating this input_dict (and save data copying along the way).

File: rllib/agents/ppo/appo_tf_policy.py [hunk 5]
```diff
@@ -444,4 +428,5 @@ def setup_late_mixins(policy: Policy, obs_space: gym.spaces.Space,
         LearningRateSchedule, KLCoeffMixin, TargetNetworkMixin,
         ValueNetworkMixin
     ],
+    view_requirements_fn=view_requirements_fn,
```
Comment (sven1977) [inline rllib/agents/ppo/appo_tf_policy.py:431]: We could actually move this into `before_loss_init` now that we have this for all policies. Then, we really don't need the view_requirements_fn in the policy templates/EagerPolicy.

File: rllib/agents/ppo/ppo_tf_policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo_tf_policy.py:308]: see above: We want the input dict to be determined by what the model needs as inputs.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 6]
```diff
@@ -73,13 +75,12 @@ def add_init_obs(self, episode_id: EpisodeID, agent_index: int,
...
                 single_row={
                     SampleBatch.OBS: init_obs,
-                    SampleBatch.EPS_ID: episode_id,
                     SampleBatch.AGENT_INDEX: agent_index,
                     "env_id": env_id,
                     "t": t,
                 })
         self.buffers[SampleBatch.OBS].append(init_obs)
-        self.buffers[SampleBatch.EPS_ID].append(episode_id)
+        self.episode_id = episode_id
```
Comment (sven1977) [inline rllib/evaluation/collectors/simple_list_collector.py:83]: Don't have to "collect" these. They are always the same for the same agent anyways.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 7]
```diff
@@ -97,15 +98,21 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> \
...
 
         for k, v in values.items():
             if k not in self.buffers:
                 self._build_buffers(single_row=values)
             self.buffers[k].append(v)
         self.count += 1
 
-    def build(self, view_requirements: Dict[str, ViewRequirement]) -> \
-            SampleBatch:
+    def build(self, view_requirements: Dict[str, ViewRequirement],
```
Comment (sven1977) [inline rllib/evaluation/collectors/simple_list_collector.py:113]: Maybe rename this to `model_view_requirements` for clarity.

Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:113]: Also: Model.inference_view_requirements -> Model.view_requirements.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 8]
```diff
@@ -200,7 +211,7 @@ def _build_buffers(self, single_row: Dict[str, TensorType]) -> None:
             ] else 0)
             # Python primitive or dict (e.g. INFOs).
             if isinstance(data, (int, float, bool, str, dict)):
-                self.buffers[col] = [0 for _ in range(shift)]
+                self.buffers[col] = [data for _ in range(shift)]
```
Comment (sven1977) [inline rllib/evaluation/collectors/simple_list_collector.py:214]: Important for custom initial state values. Cannot assume always 0 here.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 9]
```diff
@@ -251,13 +262,12 @@ def add_postprocessed_batch_for_training(
...
         """
         for view_col, data in batch.items():
-            # TODO(ekl) how do we handle this for policies that don't extend
-            # Torch / TF Policy template (no inference of view reqs)?
-            # Skip columns that are not used for training.
-            # if view_col not in view_requirements or \
-            #         not view_requirements[view_col].used_for_training:
-            #     continue
-            self.buffers[view_col].extend(data)
+            # 1) If col is not in view_requirements, we must have a direct
```
Comment (sven1977) [inline rllib/evaluation/collectors/simple_list_collector.py:265]: I think this solves it:
- if - after postprocessing - some column is not in the view-reqs, we must deal with a base-Policy child (w/o auto-view-requirement handling) -> leave as is
- if we do have it in the view reqs AND used_for_training is False -> we must have gone through auto-detection, so it's save to remove it here (this column won't be needed for training).

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 10]
```diff
@@ -442,17 +452,22 @@ def get_inference_input_dict(self, policy_id: PolicyID) -> \
...
-                                      SampleBatch.EPS_ID,
                                       SampleBatch.AGENT_INDEX] else 0)
             data_list = []
             for k in keys:
-                if data_col not in buffers[k]:
-                    self.agent_collectors[k]._build_buffers({
-                        data_col: view_req.space.sample()
-                    })
-                data_list.append(buffers[k][data_col][time_indices])
+                if data_col == SampleBatch.EPS_ID:
```
Comment (sven1977) [inline rllib/evaluation/collectors/simple_list_collector.py:460]: same as above: episode_id is always the same within one agent's collector. No need to collect an extra buffer here.

File: rllib/evaluation/rollout_worker.py [hunk 11]
```diff
@@ -177,6 +177,7 @@ def __init__(
             fake_sampler: bool = False,
             spaces: Optional[Dict[PolicyID, Tuple[gym.spaces.Space,
                                                   gym.spaces.Space]]] = None,
+            _use_trajectory_view_api: bool = True,
```
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:180]: Pass this in explicitly now into RolloutWorker (was derived from `policy_config` before, which is problematic as this could be a partial config dict)

File: rllib/policy/dynamic_tf_policy.py [hunk 12]
```diff
@@ -292,14 +290,6 @@ def __init__(
                     action_distribution=action_dist,
                     timestep=timestep,
                     explore=explore)
-            if self.config["_use_trajectory_view_api"]:
```
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py:295]: Not needed.

File: rllib/policy/dynamic_tf_policy.py [hunk 13]
```diff
@@ -417,42 +407,37 @@ def _get_input_dict_and_dummy_batch(self, view_requirements,
                 input_dict/dummy_batch tuple.
         """
         input_dict = {}
-        dummy_batch = {}
         for view_col, view_req in view_requirements.items():
             # Point state_in to the already existing self._state_inputs.
             mo = re.match("state_in_(\d+)", view_col)
             if mo is not None:
                 input_dict[view_col] = self._state_inputs[int(mo.group(1))]
-                dummy_batch[view_col] = np.zeros_like(
```
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py:426]: Better do all these in one call below.

File: rllib/policy/dynamic_tf_policy.py [hunk 14]
```diff
@@ -509,6 +494,8 @@ def fake_array(tensor):
         batch_for_postproc = UsageTrackingDict(sb)
         batch_for_postproc.count = sb.count
         logger.info("Testing `postprocess_trajectory` w/ dummy batch.")
+        self.exploration.postprocess_trajectory(
```
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py:497]: Have to also call exploration's postprocessing (may access fields in the batch we need to track; e.g. curiosity).

File: rllib/policy/eager_tf_policy.py [hunk 15]
```diff
@@ -293,10 +283,16 @@ def __init__(self, observation_space, action_space, config):
             #  Just like torch Policy does.
             self._optimizer = optimizers[0] if optimizers else None
 
+            self._initialize_loss_from_dummy_batch(
```
Comment (sven1977) [inline rllib/policy/eager_tf_policy.py:286]: Moved this here for consistency (same behavior as TorchPolicy). Also fixes a problem with curiosity where we do need the optimizer before loss init.

File: rllib/policy/policy.py [hunk 16]
```diff
@@ -700,27 +715,33 @@ def _get_dummy_batch_from_view_requirements(
...
+                    ret[view_col] = np.zeros_like(
+                        [view_req.space.sample() for _ in range(batch_size)])
+                else:
+                    ret[view_col] = [view_req.space for _ in range(batch_size)]
+
         return SampleBatch(ret)
 
     def _update_model_inference_view_requirements_from_init_state(self):
-        """Uses this Model's initial state to auto-add necessary ViewReqs.
+        """Uses Model's (or this Policy's) init state to add needed ViewReqs.
```
Comment (sven1977) [inline rllib/policy/policy.py:727]: Make this more robust against Policies that don't have a model, but do return something from `get_initial_state()`.

File: rllib/tests/test_multi_agent_env.py [hunk 17]
```diff
@@ -321,21 +323,31 @@ def compute_actions(self,
...
-                            t=t,
-                            eps_id=rollout_id,  # new id for each rollout
-                            obs=obs_batch[0],
-                            actions=0,
-                            rewards=0,
-                            dones=t == 4,
-                            infos={},
-                            new_obs=obs_batch[0])
-                    batch = builder.build_and_reset(episode=None)
+                    env_id = episodes[0].env_id
```
Comment (sven1977) [inline rllib/tests/test_multi_agent_env.py:326]: Changed test case to use new SampleCollector.

File: rllib/agents/ppo/ppo_tf_policy.py [hunk 18]
```diff
@@ -193,13 +193,22 @@ def postprocess_ppo_gae(
...
     # Trajectory has been truncated -> last r=VF estimate of last obs.
     else:
-        next_state = []
-        for i in range(policy.num_state_tensors()):
-            next_state.append(sample_batch["state_out_{}".format(i)][-1])
-        last_r = policy._value(sample_batch[SampleBatch.NEXT_OBS][-1],
-                               sample_batch[SampleBatch.ACTIONS][-1],
-                               sample_batch[SampleBatch.REWARDS][-1],
-                               *next_state)
+        # Input dict is provided to us automatically via the Model's
```
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/agents/ppo/ppo_tf_policy.py:196]: Is the comment inaccurate now?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 19]
```diff
@@ -119,28 +126,28 @@ def build(self, view_requirements: Dict[str, ViewRequirement]) -> \
                 requirements dict needed to build the SampleBatch from the raw
                 buffers (which may have data shifts as well as mappings from
                 view-col to data-col in them).
+            inference_view_requirements (Dict[str, ViewRequirement]: The view
+                requirements dict needed to build an input dict for a ModelV2
+                forward call.
```
Comment (ericl) [inline rllib/evaluation/collectors/simple_list_collector.py:131]: This argument doesn't seem to be used, can we remove it?

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Looks good, but please resolve comments before merging.

**Comment by sven1977** [comment]
:+1: Will do.

File: rllib/agents/ppo/ppo_tf_policy.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/ppo_tf_policy.py:196]: No, this is still valid.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 19 above)
Comment (sven1977) [reply rllib/evaluation/collectors/simple_list_collector.py:131]: done.
